File: mysql-async/src/main/scala/com/github/mauricio/async/db/mysql/decoder/HandshakeV10Decoder.scala
/*
 * Copyright 2013 Maurício Linhares
 *
 * Maurício Linhares licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package com.github.mauricio.async.db.mysql.decoder

import io.netty.buffer.ByteBuf
import com.github.mauricio.async.db.mysql.message.server.{HandshakeMessage, ServerMessage}
import com.github.mauricio.async.db.util.{Log, ByteBufferUtils}
import java.nio.charset.Charset

object HandshakeV10Decoder {
  final val log = Log.get[HandshakeV10Decoder]
  final val SeedSize = 8
  final val SeedComplementSize = 12
  final val Padding = 10
}

class HandshakeV10Decoder(charset: Charset) extends MessageDecoder {

  import HandshakeV10Decoder._

  def decode(buffer: ByteBuf): ServerMessage = {

    val serverVersion = ByteBufferUtils.readCString(buffer, charset)
    val connectionId = buffer.readInt()

    var seed = new Array[Byte]( SeedSize + SeedComplementSize )
    buffer.readBytes(seed, 0, SeedSize)

    buffer.readByte()

    var serverCapabilityFlags: Int = buffer.readShort()

    val characterSet = buffer.readByte() & 0xff
    val statusFlags = buffer.readShort()

    serverCapabilityFlags += 65536 * buffer.readShort().asInstanceOf[Int]

    val authPluginDataLength = buffer.readUnsignedByte()
    var authenticationMethod: Option[String] = None

    if (authPluginDataLength > 0) {
      buffer.readerIndex(buffer.readerIndex() + Padding)
      buffer.readBytes(seed, SeedSize, SeedComplementSize)
      buffer.readByte()
      authenticationMethod = Some(ByteBufferUtils.readUntilEOF(buffer, charset))
    }

    new HandshakeMessage(
      serverVersion,
      connectionId,
      seed,
      serverCapabilityFlags,
      characterSet = Some(characterSet),
      statusFlags = Some(statusFlags),
      authenticationMethod = authenticationMethod
    )
  }

}

File: mysql-async/src/main/java/com/github/jasync/sql/db/mysql/decoder/HandshakeV10Decoder.kt
package com.github.jasync.sql.db.mysql.decoder

import com.github.jasync.sql.db.mysql.encoder.auth.AuthenticationMethod
import com.github.jasync.sql.db.mysql.message.server.HandshakeMessage
import com.github.jasync.sql.db.mysql.message.server.ServerMessage
import com.github.jasync.sql.db.mysql.util.MySQLIO.CLIENT_PLUGIN_AUTH
import com.github.jasync.sql.db.mysql.util.MySQLIO.CLIENT_SECURE_CONNECTION
import com.github.jasync.sql.db.util.readCString
import com.github.jasync.sql.db.util.readUntilEOF
import io.netty.buffer.ByteBuf
import io.netty.util.CharsetUtil
import kotlin.experimental.and
import mu.KotlinLogging

private val logger = KotlinLogging.logger {}

class HandshakeV10Decoder : MessageDecoder {

    companion object {
        private const val SeedSize = 8
        private const val SeedComplementSize = 12
        private const val Padding = 10
        private val ASCII = CharsetUtil.US_ASCII
    }

    override fun decode(buffer: ByteBuf): ServerMessage {

        val serverVersion = buffer.readCString(ASCII)
        val connectionId = buffer.readUnsignedInt()

        val seed = ByteArray(SeedSize + SeedComplementSize)
        buffer.readBytes(seed, 0, SeedSize)

        buffer.readByte() // filler

        // read capability flags (lower 2 bytes)
        var serverCapabilityFlags = buffer.readUnsignedShort()

        /* New protocol , 16 bytes to describe server characteristics */
        // read character set (1 byte)
        val characterSet = buffer.readByte() and 0xff.toByte()
        // read status flags (2 bytes)
        val statusFlags = buffer.readUnsignedShort()

        // read capability flags (upper 2 bytes)
        serverCapabilityFlags = serverCapabilityFlags or (buffer.readUnsignedShort() shl 16)

        var authPluginDataLength = 0.toByte()
        var authenticationMethod = AuthenticationMethod.Native

        if ((serverCapabilityFlags and CLIENT_PLUGIN_AUTH) != 0) {
            // read length of auth-plugin-data (1 byte)
            authPluginDataLength = buffer.readByte() and 0xff.toByte()
        } else {
            // read filler (<00>)
            buffer.readByte()
        }

        // next 10 bytes are reserved (all <00>)
        buffer.readerIndex(buffer.readerIndex() + Padding)

        logger.debug("Auth plugin data length was $authPluginDataLength")

        if ((serverCapabilityFlags and CLIENT_SECURE_CONNECTION) != 0) {
            val complement = if (authPluginDataLength > 0) {
                authPluginDataLength - 1 - SeedSize
            } else {
                SeedComplementSize
            }

            buffer.readBytes(seed, SeedSize, complement)
            buffer.readByte()
        }

        if ((serverCapabilityFlags and CLIENT_PLUGIN_AUTH) != 0) {
            authenticationMethod = buffer.readUntilEOF(ASCII)
        }

        val message = HandshakeMessage(
            serverVersion,
            connectionId,
            seed,
            serverCapabilityFlags,
            characterSet = characterSet.toInt(),
            statusFlags = statusFlags,
            authenticationMethod = authenticationMethod
        )

        logger.debug { "handshake message was $message" }

        return message
    }
}
